src/metabase/db/custom_migrations/metrics_v2.clj

```diff
@@ -0,0 +1,158 @@
+  of the corresponding entities in `id->entity`.
+  `expr` is an MBQL aggregation expression or a part of it.
+  The mapping is from metric IDs to v2 metric cards."
+  [expr id->entity]
+  (if-not (vector? expr)
+    expr
+    (if-let [id (metric-ref->id expr)]
+      (assoc expr 1 (-> id id->entity :id))
+      (mapv #(replace-metric-refs % id->entity) expr))))
+
+(defn- rewrite-metric-consuming-query
+  [query metric-id->metric-card]
+  (if (contains? query :source-query)
+    (update query :source-query rewrite-metric-consuming-query metric-id->metric-card)
+    (let [aggregation (:aggregation query)
+          rewritten (replace-metric-refs aggregation metric-id->metric-card)]
+      (cond-> query
+        (not= rewritten aggregation) (assoc :aggregation rewritten)))))
+
+(defn- rewrite-metric-consuming-card
+  "Rewrite the question `card` replacing references to v1 metrics with references
+  to the corresponding v2 metric question as specified by the mapping `metric-id->metric-card`."
+  [card metric-id->metric-card]
+  (let [dataset-query (json/parse-string (:dataset_query card) true)
+        inner-query (:query dataset-query)
+        rewritten (rewrite-metric-consuming-query inner-query metric-id->metric-card)]
+    (when (not= rewritten inner-query)
+      (-> card
+          (assoc :dataset_query (-> dataset-query
+                                    (assoc :query rewritten)
+                                    json/generate-string))
+          (dissoc :updated_at)))))
+
+(defn migrate-up
+  "Migrate metrics and the cards consuming them to metrics v2. This involves
+  1. creating the migration collection,
+  2. creating a v2 metric card for each v1 metric in the migration collection,
+  3. making a backup copy of the dataset_query of each metric consuming card,
+  4. rewriting the metric consuming cards to use the migrated v2 metric cards."
+  []
+  (when (t2/exists? :metric)
+    (let [metric-v2-coll (get-metric-migration-collection {:create? true})
+          metric-id->metric-card (into {}
+                                       (map (juxt :id #(create-metric-v2 % metric-v2-coll)))
+                                       (t2/query {:select [:m.* [:t.db_id :database_id]]
+                                                  :from [[:metric :m]]
+                                                  :left-join [[:metabase_table :t] [:= :t.id :m.table_id]]}))
+          metric-consuming-cards (t2/select :report_card {:where [:like [:lower :dataset_query] "%[\"metric\",%"]})]
+      (doseq [card metric-consuming-cards
+              :let [rewritten (rewrite-metric-consuming-card card metric-id->metric-card)]
+              :when (and rewritten (not= card rewritten))]
+        (t2/update! :report_card
+                    (:id card)
+                    (-> rewritten
+                        (assoc :dataset_query_metrics_v2_migration_backup (:dataset_query card)
+                               :updated_at (Instant/now))))))))
```

Comment:
I think it's usually better to use `current_timestamp` or `now()` (or `now(3)` for MySQL/MariaDB to get millisecond precision, altho that's probably not important in this case) so you don't have to worry about weirdness happening here with timezone conversion if this is a `datetime` or `timestamp [without time zone]` column but I guess it doesn't really matter that much

Reply:
Yes, in fact, I'm not 100% sure that updating this column is what everybody would wish.

Reply:
I've removed this occurrence altogether, but there is another one. Is this important? As long as setting the timestamp works, I think we are fine.